fv3config/filesystem.py

```diff
@@ -61,19 +70,31 @@ def is_local_path(location):
     return _get_protocol_prefix(location) == ""
 
 
-def _put_directory(local_source_dir, dest_dir, fs=None):
+def put_directory(
+    local_source_dir: str,
+    dest_dir: str,
+    fs: fsspec.AbstractFileSystem = None,
+    executor: Executor = None,
+):
     """Copy the contents of a local directory to a local or remote directory.
     """
     if fs is None:
         fs = get_fs(dest_dir)
+    if executor is None:
+        executor = ThreadPoolExecutor()
+        manage_threads = True
+    else:
+        manage_threads = False
     for token in os.listdir(local_source_dir):
         source = os.path.join(os.path.abspath(local_source_dir), token)
         dest = os.path.join(dest_dir, token)
         if os.path.isdir(source):
-            fs.makedirs(dest, exist_ok=True)
-            _put_directory(source, dest, fs)
+            fsspec_backoff(fs.makedirs)(dest, exist_ok=True)  # must be blocking call
+            put_directory(source, dest, fs=fs, executor=executor)
         else:
-            fs.put(source, dest)
+            executor.submit(fsspec_backoff(fs.put), source, dest)
```

Comment:
Was there a particular reason to have our own recursive function instead of just calling `fs.put(local_source_dir, dest_dir, recursive=True)`?

Fine to leave as is--and I'm actually not sure if the multi-threading would work if it wasn't written the way it is now.

Reply:
I initially wrote here "It allows the option of controlling the threader in the calling context - you can initialize a ThreadPoolExecutor (or any other executor) in the calling context, and pass it in on the initial call." but I misunderstood your question.

When I wrote this initially (I think I wrote it) I didn't know recursive=True was an option. Probably should have used that initially. At this point, to get threaded copy, we'll need our own recursive function (unless fsspec does threading internally, which it doesn't appear that it does, looking at the source).

Reply:
Thanks!